### DISCUSSION TITLE
Events passed to named actions are not strongly typed.

### DISCUSSION THREAD

**Discussion by UnitRoot22** [statelyai/xstate#4686] [Q&A] [answered]
I am trying to migrate a project of mine from xstate v4 to v5. 
I understand typegen is not currently supported and that was what enabled named actions to be typed correctly.
Here is a trivial example of what I am trying to accomplish

```typescript
import { assign, setup, fromPromise, assertEvent } from "xstate";

const basicFetchMachine = setup({
  types: {
    context: {} as {
      currentYear: string;
    },
  },
  actions: {
    assignCurrentYearToContext: assign(({ event }) => {
      return {
        // event.ouput should be strongly typed to <string>, but it is not
        currentYear: event.output,
      };
    }),
  },
  actors: {
    fetchCurrentYear: fromPromise<string>(async () => {
      return Promise.resolve("2077");
    }),
  },
}).createMachine({
  initial: "fetchingUntyped",
  context: {
    currentYear: "",
  },
  states: {
    fetchingUntyped: {
      invoke: {
        src: "fetchCurrentYear",
        onDone: {
          actions: "assignCurrentYearToContext",
        },
      },
    },
    fetchingStronglyTyped: {
      invoke: {
        src: "fetchCurrentYear",
        onDone: {
          actions: assign(({ event }) => {
            return {
              // event.output is correctly typed to <string>
              currentYear: event.output,
            };
          }),
        },
      },
    },
  },
});
```
The two comments indicate where strong typing is available vs not. 
I have seen the discussions here: https://github.com/statelyai/xstate/discussions/4556
But, even doing something like this below, does not fix the issue
```typescript
        onDone: [
          {
            actions: {
              type: "assignCurrentYearToContext",
            },
          },
        ],
```
In v5, is there a way to make named actions strongly typed other than just being forced to inline them ?

**Comment by davidkpiano** [comment, marked as answer]
We strongly encourage you to use **parameterized actions** instead of relying on the event. This allows greater flexibility and better type inference for actions:

```ts
import { assign, setup, fromPromise, assertEvent } from "xstate";

const basicFetchMachine = setup({
  types: {
    context: {} as {
      currentYear: string;
    },
  },
  actions: {
    // Use params instead of event
    assignCurrentYearToContext: assign((_, year: string) => {
      return {
        currentYear: year,
      };
    }),
  },
  actors: {
    fetchCurrentYear: fromPromise<string>(async () => {
      return Promise.resolve("2077");
    }),
  },
}).createMachine({
  initial: "fetchingUntyped",
  context: {
    currentYear: "",
  },
  states: {
    fetchingUntyped: {
      invoke: {
        src: "fetchCurrentYear",
        onDone: {
          actions: {
            type: "assignCurrentYearToContext",
            // Send event.output (strongly typed) as params (strongly typed)
            params: ({ event }) => event.output
          }
        },
      },
    },
    fetchingStronglyTyped: {
      invoke: {
        src: "fetchCurrentYear",
        onDone: {
          actions: assign(({ event }) => {
            return {
              // event.output is correctly typed to <string>
              currentYear: event.output,
            };
          }),
        },
      },
    },
  },
});
```

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgQwM6uAcwHYBo6oCmMArmPgGZQQgAK1IwR+aRsAogG6HbwC+cKjTgAiAB6oYyGIREBuALAAoZQGMI2SXABGaYKoBixVQAsAsslPBshOAF4CxMgAoEyuHBgBPMIVQAuRHcPOHVeQjEYQIQBNCClEJDVEigoHhgATUJkKEDJKGtMRQSQvlxgsuDLGGANAPjE1ixsAGEUtN4snIAVCBaNGUjAppxnZwB9fC9s3IIYAuxMAEp7AD4GxLg00ihsDc3Q9vSu2emc8pLEvmKrpYuPSpLq6Hq3S8FjEzbU45nAoToDCYhAAPPlCqtnGgvNhVHBnCs7Os3gdtik9vQaMCAHRpVAQAA23GcIgATAAGADslJESxupTuFQufCW2NUaWkhAsVhsrmC1mANWQBMCIgon0KAFVeD5CAATET3UIDCJRfaHH6dP6iRVM4KSTmvYIecUwHmYaXeXxy6LGkLWTgQADWhFt70SqCgqlFptM3w6mRmuvdIQ0ABENK71QdqrVNG6DgcrVGRCNWkctT0+irIsHE5swDlkCB6q44IRuLw4Cy1uXKzBsRASDAwM27Qc+O3Skqrj3q33fSZCgBleYaTAErzdWU26NwB3OqMoxOe72iQf+345POJ8ORhP5lCqGp1YboZpjJAV9LVxHIruopy7Of5gD0r7r6Ubzdb8CYyp+Y9J08GdPAgOAwXmCEH0TZJNUDHJAmvXhvxbZs+3za4YIeRkQweDD+ztR58OUFk5CAA
